### PR TITLE
API-7369: Add support for PAT Authorization

### PR DESCRIPTION
**Pull request by czeczot**
* Added TokenType - bearer/basic
* Authorization header created from tokenType and access token
* Added test cases for basic / bearer / invalid token type

### CODE REVIEW

File: authorization/types.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package authorization
+
+// TokenType represents Bearer or Basic authentication scheme.
+type TokenType int
+
+// Possible values of TokenType.
```
Comment (litwinow) [inline authorization/types.go:6]:
```suggestion
// Supported values of TokenType.
```

File: authorization/types.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package authorization
+
+// TokenType represents Bearer or Basic authentication scheme.
```
Comment (litwinow) [inline authorization/types.go:3]:
```suggestion
// TokenType represents authentication scheme.
```

File: changelog.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Changelog
+
+
+### [v2.1.0]
+
+Authorization
+
+* added TokenType to Token in order to enable authorization by PAT
```
Comment (litwinow) [inline changelog.md:8]:
```suggestion
* Added Type to authorization.Token in order to support different authentication schemes (`Bearer` and `Basic`).
```

File: go.sum [hunk 4]
```diff
@@ -0,0 +1 @@
+github.com/livechat/lc-sdk-go v1.0.0 h1:sMIkj3p3emAs9nS4TIjhwhsArxYnPcB60aZgRPt6I2I=
```
Comment (litwinow) [inline go.sum]: pls remove

File: internal/web_api.go [hunk 5]
```diff
@@ -178,3 +181,10 @@ func DefaultHTTPRequestGenerator(name string) HTTPRequestGenerator {
 		return http.NewRequest("POST", url, nil)
 	}
 }
+
+func (a *api) tokenTypeToString(tokenType authorization.TokenType) string {
+	if tokenType == authorization.BasicToken {
+		return "Basic"
+	}
+	return "Bearer"
+}
```
Comment (kogierman) [inline internal/web_api.go:190]: please make it a `String()` method with `TokenType` as receiver, then you can simply pass it to `fmt.Sprintf`

File: agent/api_test.go [hunk 6]
```diff
@@ -351,6 +365,42 @@ func createMockedErrorResponder(t *testing.T, method string) func(req *http.Requ
...
+
+		if tokenType == authorization.BasicToken && authHeader != "Basic access_token" {
+			t.Errorf("Invalid Authorization header: %s", authHeader)
+			return createServerError("Invalid Authorization")
+		}
+
+		if tokenType == authorization.BearerToken && authHeader != "Bearer access_token" {
+			t.Errorf("Invalid Authorization header: %s", authHeader)
+			return createServerError("Invalid Authorization")
+		}
```
Comment (litwinow) [inline agent/api_test.go:394]: you could extract it into functor and pass it as argument to `createAuthorizationHeaderChecker` - `createAuthorizationHeaderChecker` doesn't need to have knowledge of all supported test cases

Comment (litwinow) [reply agent/api_test.go:394]: Also pls have a look at `TestAuthorIDHeader` test - it does very similar stuff to what you want to achieve

**Review by litwinow — CHANGES_REQUESTED** [review]

**Review by kogierman — CHANGES_REQUESTED** [review]
+ the stuff that Paweł mentioned

File: agent/api_test.go [hunk 7]
```diff
@@ -26,12 +26,26 @@ func NewTestClient(fn roundTripFunc) *http.Client {
...
+}
+
+func stubBasicTokenGetter() *authorization.Token {
+	return stubGenericTokenGetter(authorization.BasicToken)
+}
+
+func stubInvalidTokenGetter() *authorization.Token {
+	invalidTokenType := 2020
+	return stubGenericTokenGetter(authorization.TokenType(invalidTokenType))
+}
```
[the remark this answers is not in the document]
Comment (czeczot) [reply agent/api_test.go:40]: but NewApi method accepts func without parameters so passing stubGenericTokenGetter won't work

[the remark this answers is not in the document]
Comment (litwinow) [reply agent/api_test.go:40]: `stubGenericTokenGetter` can return function

**Review by kogierman — APPROVED** [review]

**Review by litwinow — APPROVED** [review]

**Pull request merged**